### PR TITLE
Improve the performance of replication

### PR DESCRIPTION
**Pull request by ywk253100**
Improve the performance of replication by introducing a new API to check whether the blob can be mounted directly

Signed-off-by: Wenkai Yin <yinw@vmware.com>

### CODE REVIEW

**Comment by ywk253100** [comment]
Trying to replicate the same image goang:1.15.6(300M) from 50 different repositories and get the following test result:
|  | Jobservice Woker Count | Time Spent |
|--|--|--|
| Without Improvement | 10 | 5min32s |
| With Improvement | 10 | 42s |
| Without Improvement | 1 | 15min35s |
| With Improvement | 1 | 77s |

File: src/pkg/reg/adapter/harbor/v2/adapter.go [hunk 1]
```diff
@@ -129,3 +129,14 @@ func (a *adapter) listArtifacts(repository string, filters []*model.Filter) ([]*
 	}
 	return filter.DoFilterArtifacts(artifacts, filters)
 }
+
+func (a *adapter) CanBeMount(digest string) (bool, string, error) {
+	repository, err := a.client.getRepositoryByBlobDigest(digest)
```
Comment (heww) [inline src/pkg/reg/adapter/harbor/v2/adapter.go:134]: Do we need to process the 404 here for the previous version of harbor?

File: api/v2.0/swagger.yaml [hunk 2]
```diff
@@ -384,6 +384,43 @@ paths:
...
+              type: string
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/Repository'
+        '400':
+          $ref: '#/responses/400'
+        '401':
+          $ref: '#/responses/401'
+        '403':
```
Comment (reasonerjt) [inline api/v2.0/swagger.yaml:418]: In what case it returns 403 or 404?

File: src/pkg/reg/adapter/harbor/v2/adapter.go (hunk 1 above)
Comment (ywk253100) [reply src/pkg/reg/adapter/harbor/v2/adapter.go:135]: updated

File: src/common/models/repo.go [hunk 3]
```diff
@@ -38,6 +45,48 @@ type RepoRecord struct {
 	UpdateTime   time.Time `orm:"column(update_time);auto_now" json:"update_time"`
 }
 
+// FilterByBlobDigest filters the repositories by the blob digest
+func (r *RepoRecord) FilterByBlobDigest(ctx context.Context, qs orm.QuerySeter, key string, value interface{}) orm.QuerySeter {
+	digest, ok := value.(string)
```
Comment (steven-zou) [inline src/common/models/repo.go:50]: If you only accept a `string` as arg, why not directly declare the value to a `string` type? The caller needs to guarantee that the passing parameter should be a string value.

**Review by steven-zou — APPROVED** [review]
LGTM with a minor comment

File: src/common/models/repo.go (hunk 3 above)
Comment (ywk253100) [reply src/common/models/repo.go:50]: The method `FilterByXXX` is an interface defined for the ORM query, it isn't only for the repository

**Review by heww — APPROVED** [review]

**Pull request merged**